### PR TITLE
Fix FileSystemAclExtensions.Create when passing a null FileSecurity

### LINKED ISSUE #57768: FileSystemAclExtensions.Create broken in multiple ways
On .NET Framework, this:
```C#
using System;
using System.IO;
using System.Security.AccessControl;

internal class Program
{
    static void Main()
    {
        string path = @"tmp.txt";
        using (FileStream fs = new FileStream(path, FileMode.Create, FileSystemRights.AppendData, FileShare.None, 1, FileOptions.None, null))
        {
            for (int i = 0; i < 26; i++)
            {
                fs.WriteByte((byte)('a' + i));
                fs.Position = 0;
            }
        }

        const string Expected = "abcdefghijklmnopqrstuvwxyz";
        string text = File.ReadAllText(path);
        Console.WriteLine(text == Expected ?
            "Success" :
            $"Expected {Expected}, got {text}");
    }
}
```
produces:
```
Success
```

On .NET 5/6, you should be able to use FileStreamAclExtensions.Create in the exact same way (its whole purpose for existing is to provide the exact same functionality on core):
```C#
using System;
using System.IO;
using System.Security.AccessControl;

internal class Program
{
    static void Main()
    {
        string path = @"tmp.txt";
        using (FileStream fs = FileSystemAclExtensions.Create(new FileInfo(path), FileMode.Create, FileSystemRights.AppendData, FileShare.None, 1, FileOptions.None, null))
        {
            for (int i = 0; i < 26; i++)
            {
                fs.WriteByte((byte)('a' + i));
                fs.Position = 0;
            }
        }

        const string Expected = "abcdefghijklmnopqrstuvwxyz";
        string text = File.ReadAllText(path);
        Console.WriteLine(text == Expected ?
            "Success" :
            $"Expected {Expected}, got {text}");
    }
}
```
but that blows up with an ArgumentNullException:
```
Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'fileSecurity')
   at System.IO.FileSystemAclExtensions.Create(FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, Int32 bufferSize, FileOptions options, FileSecurity fileSecurity)
   at Program.Main()
```

If you then fix that to allow null, it then fails deeper with how it's mapping to FileStream validation of FileAccess:
```
Unhandled exception. System.ArgumentOutOfRangeException: Enum value was out of legal range. (Parameter 'access')
   at System.IO.FileStream.ValidateHandle(SafeFileHandle handle, FileAccess access, Int32 bufferSize) in D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\FileStream.cs:line 75
   at System.IO.FileStream.ValidateHandle(SafeFileHandle handle, FileAccess access, Int32 bufferSize, Boolean isAsync) in D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\FileStream.cs:line 89
   at System.IO.FileStream..ctor(SafeFileHandle handle, FileAccess access, Int32 bufferSize, Boolean isAsync) in D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\IO\FileStream.cs:line 115
   at System.IO.FileSystemAclExtensions.Create(FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, Int32 bufferSize, FileOptions options, FileSecurity fileSecurity) in D:\repos\runtime\src\libraries\System.IO.FileSystem.AccessControl\src\System\IO\FileSystemAclExtensions.cs:line 204
   at Program.Main()
```

### PR DESCRIPTION
**Pull request by carlossanlop**
Fixes https://github.com/dotnet/runtime/issues/57768

Fixes in this PR:

- The ACL extension method `FileStream.Create` should accept a `null` `FileSecurity` parameter so the behavior can match .NET Framework. The ref definition had to be changed to a nullable parameter. @bartonjs @GrabYourPitchforks @stephentoub  Do I need to go through API review to request this change or can we validate it here? ~This will be marked as a breaking change because an exception case was removed~.

- `FileStream` throws `ArgumentOutOfRangeException` if the passed `FileAccess` is not valid. In .NET Framework, this was not a problem because all the ACL logic was contained within the `FileStream` constructor, so it was possible to simply pass the `FileSystemRights` enum value casted as a `FileAccess` enum value to the p/invoke. Here, the ACL logic is in an extension method, and we don't want to do such casting. To address this, I am now _mapping_ all the writeable `FileSystemRights` values to a `FileAccess` write or read value. I left a detailed explanation [here](https://github.com/dotnet/runtime/issues/57768#issuecomment-923503098) of what we were doing before.

- Additional bug fix: The `FileShare.Inheritable` bit needs to be saved in the `SECURITY_ATTRIBUTES` struct (we are already doing this), but should not be passed to the `CreateFile` P/Invoke (we were not doing this). Otherwise, the P/Invoke causes an `IOException` to be thrown with the message `The parameter is incorrect`.

- Added unit tests to verify all the write combinations when the `FileSecurity` argument is `null`, which also includes the code example in the original issue.

### CODE REVIEW

**Comment by bartonjs** [comment]
> The ACL extension method FileStream.Create should accept a null FileSecurity parameter so the behavior can match .NET Framework. ... API Review ... ?

Moving from not allowing null to allowing null is fine.  I don't think as reviewers we have the muscles developed to be good at inquiring about the state of question marks, and that direction isn't a breaking change.

> This will be marked as a breaking change because an exception case was removed.

This the poster child for a "visible behavioral change" that isn't a breaking change.  (ArgumentNullException means the problem was in their code.  So if we allow null now and they didn't want to, that error is still in their code)

https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/breaking-change-rules.md (Exceptions, Allowed, last bullet)

**Review by jeffhandley — APPROVED** [review]
Looks good to me. Let's get another review from @adamsitnik or @Jozkee though.

**Review by jozkee — COMMENTED** [review]
Left some suggestions for you to consider.

**Review by adamsitnik — COMMENTED** [review]
`GetFileStreamFileAccess` should not return `0`, as `FileStream` is simply going to throw. Please see my comments.

Do we have a unit test that ensures that `FileSystemRights.AppendData` is working the way it was shown in #57768?

File: src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs [hunk 1]
```diff
@@ -380,7 +378,8 @@ private void CreateFileWithSecurity(FileInfo info, FileMode mode, FileSystemRigh
             }
```
Comment (carlossanlop) [inline src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:378]: Do we still need to check for `PlatformDetection.IsNetFramework` or can I remove those checks?

File: src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs [hunk 2]
```diff
@@ -565,9 +586,12 @@ private void Verify_FileSecurity_CreateFile(FileMode mode, FileSystemRights righ
             Assert.True(fileInfo.Exists);
             tempRootDir.CreatedSubfiles.Add(fileInfo);
 
-            var actualFileInfo = new FileInfo(path);
-            FileSecurity actualSecurity = actualFileInfo.GetAccessControl(AccessControlSections.Access);
-            VerifyAccessSecurity(expectedSecurity, actualSecurity);
+            if (expectedSecurity != null)
+            {
+                var actualFileInfo = new FileInfo(path);
+                FileSecurity actualSecurity = actualFileInfo.GetAccessControl(AccessControlSections.Access);
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:592]: It returns a `FileSecurity` object that contains zero access rules. I added a test for that case.

File: src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs [hunk 3]
```diff
@@ -550,6 +548,29 @@ public static IEnumerable<object[]> RightsToAllow()
...
+            FileSystemRights.FullControl,
+            FileSystemRights.Modify,
+            FileSystemRights.Write
+        };
+
+        public static IEnumerable<object[]> ValidArgumentsWriteableStreamNullFileSecurity() =>
+            from mode in new[] { FileMode.Create, FileMode.CreateNew, FileMode.OpenOrCreate }
+            from rights in s_writeableFileSystemRights
+            from share in new[] { FileShare.None, FileShare.Read, FileShare.Write, FileShare.ReadWrite }
+            from bufferSize in new[] { 1, DefaultBufferSize }
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:571]: Removed.

File: src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs [hunk 4]
```diff
@@ -380,7 +378,8 @@ private void CreateFileWithSecurity(FileInfo info, FileMode mode, FileSystemRigh
             }
             else
             {
-                info.Create(mode, rights, share, bufferSize, options, security);
+                info.Create(mode, rights, share, bufferSize, options, security).Dispose();
+                Assert.True(info.Exists);
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:382]: You can create a file with specific access rules and audit rules to allow or deny permissions.

File: src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs [hunk 5]
```diff
@@ -289,13 +289,11 @@ public void FileInfo_Create_NullFileInfo()
                 CreateFileWithSecurity(info, FileMode.CreateNew, FileSystemRights.WriteData, FileShare.Delete, DefaultBufferSize, FileOptions.None, security));
         }
 
-        [Fact]
-        public void FileInfo_Create_NullFileSecurity()
+        [Theory]
+        [MemberData(nameof(ValidArgumentsWriteableStreamNullFileSecurity))]
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:293]: Done.

File: src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs [hunk 6]
```diff
@@ -261,20 +256,48 @@ public static DirectoryInfo CreateDirectory(this DirectorySecurity directorySecu
         private static FileAccess GetFileStreamFileAccess(FileSystemRights rights)
         {
             FileAccess access = 0;
-            if ((rights & FileSystemRights.ReadData) != 0 || ((int)rights & Interop.Kernel32.GenericOperations.GENERIC_READ) != 0)
+
+            if (rights == (FileSystemRights)0)
+            {
+                return access;
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:262]: Weird. The unit tests are passing, the `FileStream` constructor is not throwing. I'm investigating.

File: src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs [hunk 7]
```diff
@@ -550,6 +548,29 @@ public static IEnumerable<object[]> RightsToAllow()
...
+            FileSystemRights.TakeOwnership,
+            FileSystemRights.Synchronize,
+            FileSystemRights.FullControl,
+            FileSystemRights.Modify,
+            FileSystemRights.Write
+        };
+
+        public static IEnumerable<object[]> ValidArgumentsWriteableStreamNullFileSecurity() =>
+            from mode in new[] { FileMode.Create, FileMode.CreateNew, FileMode.OpenOrCreate }
+            from rights in s_writeableFileSystemRights
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:551]: They are disallowed.

If you run a test console app on .NET Framework:

```cs
string path = "file.txt";
FileMode mode = FileMode.CreateNew;
FileSystemRights rights = FileSystemRights.ReadAndExecute;
FileShare share = FileShare.None;
int bufferSize = 4096;
FileOptions options = FileOptions.None;

using (FileStream fs = new FileStream(path, mode, rights, share, bufferSize, options, null)){}
```

You get the following exception:
```
'Combining FileMode: CreateNew with FileSystemRights: ReadAndExecute is invalid.'

This exception was originally thrown at this call stack:
    System.IO.FileStream.Init(string, System.IO.FileMode, System.IO.FileAccess, int, bool, System.IO.FileShare, int, System.IO.FileOptions, Microsoft.Win32.Win32Native.SECURITY_ATTRIBUTES, string, bool, bool, bool) in filestream.cs
    System.IO.FileStream.FileStream(string, System.IO.FileMode, System.Security.AccessControl.FileSystemRights, System.IO.FileShare, int, System.IO.FileOptions, System.Security.AccessControl.FileSecurity) in filestream.cs
    ConsoleFramework.Program.Main() in Program.cs
```

So we do a similar verification in .NET 7:

```
 System.ArgumentException : Combining FileMode.CreateNew with FileSystemRights.ReadAndExecute is invalid.
        Stack Trace:
          C:\Users\calope\source\repos\runtime\src\libraries\System.IO.FileSystem.AccessControl\src\System\IO\FileSystemAclExtensions.cs(202,0): at System.IO.FileSystemAclExtensions.Create(FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, Int32 bufferSize, FileOptions options, FileSecurity fileSecurity)
```

I'm comparing the output of .NET Framework vs .NET 7 with all the combinations of `FileMode` and `FileSystemRights` to determine if I need to special-case some unit tests.

File: src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (carlossanlop) [reply src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:262]: Also, notice that we pass `(FileAccess)0` as an argument to `Init` in .NET Framework:
https://referencesource.microsoft.com/#mscorlib/system/io/filestream.cs,569

**Comment by carlossanlop** [comment]
Closing and resending PR to make it easier to review.

**Pull request closed without merging**